text/000-structural-and-deref.md

```diff
@@ -0,0 +1,410 @@
+- Start Date: 2018-10-05
+- RFC PR: (leave this empty)
+- Tracking Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Change `#[wasm_bindgen]` to use `structural` by default, and add a new
+attribute `final` for an opt-in to today's behavior. Once implemented then use
+`Deref` to model the class inheritance hierarchy in `web-sys` and `js-sys` to
+enable ergonomic usage of superclass methods of web types.
+
+# Motivation
+[motivation]: #motivation
+
+The initial motivation for this is outlined [RFC 3], namely that the `web-sys`
+crate provides bindings for many APIs found on the web but accessing the
+functionality of parent classes is quite cumbersome.
+
+The web makes extensive use of class inheritance hierarchies, and in `web-sys`
+right now each class gets its own `struct` type with inherent methods. These
+types implement `AsRef` between one another for subclass relationships, but it's
+quite unergonomic to actually reference the functionality! For example:
+
+```rust
+let x: &Element = ...;
+let y: &Node = x.as_ref();
+y.append_child(...);
+```
+
+or...
+
+```rust
+let x: &Element = ...;
+<Element as AsRef<Node>>::as_ref(x)
+    .append_child(...);
+```
+
+It's be much nicer if we could support this in a more first-class fashion and
+make it more ergonomic!
+
+[RFC 3]: https://github.com/rustwasm/rfcs/pull/3
+
+> **Note**: While this RFC has the same motivation as [RFC 3] it's proposing an
+> alternative solution, specifically enabled by switching by `structural` by
+> default, which is discussed in [RFC 3] but is hopefully formally outlined
+> here.
+
+# Detailed Explanation
+[detailed-explanation]: #detailed-explanation
+
+This RFC proposes using the built-in `Deref` trait to model the class hierarchy
+found on the web in `web-sys`. This also proposes changes to `#[wasm_bindgen]`
+to make using `Deref` feasible for binding arbitrary JS apis (such as those on
+NPM) with `Deref` as well.
+
+For example, `web-sys` will contain:
+
+```rust
+impl Deref for Element {
+    type Target = Node;
+
+    fn deref(&self) -> &Node { /* ... */ }
```

Comment:
Small nit: I think this should be `fn deref(&self) -> &Self::Target { /* ... */ }`

Reply:
Either way works here, `Self::Target` resolved to `Node` in this case so they're interchangeable.

Reply:
I know. But it means less duplication (and it's also more idiomatic, in my opinion).